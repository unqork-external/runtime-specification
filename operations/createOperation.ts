import { OperationTypes } from './enums/operation-types.enum'
import { OperationTypeMap } from './typemaps/operation.typemap'

/**
 * Given an OperationType, Typescript will complete the options list.
 *
 * In return, the caller receives Typescript constructed class.
 *
 * This includes preset defaults.
 *
 * In essence, this function ensures that all keys are fully present, fully typed, and ready for runtime
 * validation if necessary.
 *
 * @param type
 * @param options
 * @param name name of the operation
 * @param additional
 * @param spreadOp
 * @param additionalArgs
 * @returns - The Operation Class type
 */
export const createOperation = <OpType extends OperationTypes>(
  type: OpType,
  options: InstanceType<(typeof OperationTypeMap)[OpType]>['options'],
  additionalArgs?: AdditionalArgs,
): InstanceType<(typeof OperationTypeMap)[OpType]> => {
  const classOpType = OperationTypeMap[type]
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

interface AdditionalArgs {
  notifyImmediately: boolean
}
