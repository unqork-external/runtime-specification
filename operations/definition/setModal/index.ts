export * from './setModal.options'
export * from './setModal.model'
export * from './setModal.enum'
