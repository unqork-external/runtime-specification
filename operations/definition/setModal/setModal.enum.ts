export enum ModalEventKeys {
  MODAL_OPEN_KEY = 'modalOpenEvent',
  MODAL_CLOSE_KEY = 'modalDismissEvent',
}
