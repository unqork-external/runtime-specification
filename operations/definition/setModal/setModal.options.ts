import { Description, Example, Property, Required } from '@tsed/schema'

export class SetModalOptions {
  @Description('The key or path to the container we intend to target for an operation')
  targetKey: string

  @Required()
  @Description('modalAction refers to the event that the modal should execute.')
  @Example(`MODAL_CLOSE_KEY`, `MODAL_OPEN_KEY`)
  modalAction: string

  @Property()
  targetEventKey: string
}
