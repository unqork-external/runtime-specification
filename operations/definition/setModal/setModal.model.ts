import { DiscriminatorValue, Required } from '@tsed/schema'

import { SetModalOptions } from './setModal.options'
import { OperationTypes } from '../../enums/operation-types.enum'
import { Operation } from '../../interface/operations.interface'

@DiscriminatorValue(OperationTypes.SET_MODAL)
export class SetModalOperation extends Operation<OperationTypes.SET_MODAL> {
  public type = OperationTypes.SET_MODAL as const

  @Required()
  public options: SetModalOptions = new SetModalOptions()
}
