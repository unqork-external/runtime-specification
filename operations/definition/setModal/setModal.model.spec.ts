import { SetModalOperation } from './setModal.model'
import { generateSchemaAndValidate } from '../../../../utilities'
import { createOperation } from '../../createOperation'
import { OperationTypes } from '../../enums/operation-types.enum'

describe('setModal', () => {
  let validate, schema
  beforeAll(() => {
    const response = generateSchemaAndValidate(SetModalOperation)
    validate = response.validate
    schema = response.schema
  })

  it('should match this JSON Schema', () => {
    expect(schema).toMatchSnapshot()
  })

  it('should validate submit operation with all required keys', () => {
    const op = createOperation(OperationTypes.SET_MODAL, {
      modalAction: 'MODAL_CLOSE_KEY',
      targetEventKey: '1',
      targetKey: '1',
    })

    const isValid = validate(op)
    expect(isValid).toBeTrue()
  })

  it('should error if no modalAction is provided', () => {
    //@ts-ignore
    const op = createOperation(OperationTypes.SET_MODAL, {})

    const isValid = validate(op)
    expect(isValid).toBeFalse()
  })
})
