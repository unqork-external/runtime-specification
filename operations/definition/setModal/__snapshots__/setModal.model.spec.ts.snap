// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setModal should match this JSON Schema 1`] = `
{
  "definitions": {
    "SetModalOptions": {
      "properties": {
        "modalAction": {
          "description": "modalAction refers to the event that the modal should execute.",
          "examples": [
            "MODAL_CLOSE_KEY",
            "MODAL_OPEN_KEY",
          ],
          "minLength": 1,
          "type": "string",
        },
        "targetEventKey": {
          "type": "string",
        },
        "targetKey": {
          "description": "The key or path to the container we intend to target for an operation",
          "type": "string",
        },
      },
      "required": [
        "modalAction",
      ],
      "type": "object",
    },
  },
  "properties": {
    "name": {
      "description": "Name of the operation",
      "type": "string",
    },
    "options": {
      "$ref": "#/definitions/SetModalOptions",
    },
    "type": {
      "const": "SET_MODAL",
      "examples": [
        "SET_MODAL",
      ],
      "type": "string",
    },
  },
  "required": [
    "options",
  ],
  "type": "object",
}
`;
